0001-numerics.md

```diff
@@ -0,0 +1,93 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Serde Issue: (leave this empty)
+
+# Summary
+
+One paragraph explanation of the feature.
+
+# Motivation
+
+**TLDR**: Let's do a clean restart of number serialization fitting everyone's needs. Skip forward to *Proposal*
+
+There is the set `A` of number representations that the serialization formats use. Then there's the set `B` of number representations in rust. `B` also contains arbitrary user types (`BigNum`, `d128`, ...).
+
+Ideally we'd have a function `f(A) -> B` and `f'(B) -> A`. Obviously that makes little sense, as it would require serde to know about all possible types. So instead we have a set `S` that contains all basic rust number types and  `f(A) -> S`, `f'(S)->A` as well as `g(B) -> S`, `g'(S) -> B`.
+
+Right now we are discussing what `B` should contain and how to represent it. The problem I'm seeing is that we are discussing the memory representation instead of the abstract representation. There will never be a perfect memory representation that fits all needs (e.g. `BigNum` can represent all numbers, but requires memory allocation). We'll always have to do tradeoffs. The tradeoff decision shouldn't really be made by `serde` or an implementor, but by the final user. We could of course get there through a huge `cfg!` madness, but I have a proposal (no concrete design, just some thoughts):
+
+# Detailed design
+
+We add a new trait `NumericSerialization` that is implemented for the regular rustc types. Users can implement this trait for their own types. Also there's a type `NumericTypeInfo` that contains various things that are useful to know about the type.
+
+```rust
+struct NumericTypeInfo<T: NumericSerialization> {
+    name: &'static str,
+    // if you have to hold more than 2^64 bits, then I have bad news for you.
+    // None for infinitely sized types
+    min_bits: Option<u64>,
+    min_value: Option<&'static T>,
+    // can a value of this type be signed
+    signed: bool,
+    // can there only be integral values? (only zeros after the comma)
+    integral: bool,
+}
+
+/// `Display` bound for textual representations
+trait NumericSerialization: core::fmt::Display {
```

Comment:
should be easy enough. if `is_integral` is false, then one can either use `as_f64` and we can offer a way to split the `NumericSerialization` into two parts: one before the decimal and one after the decimal. Both of these are again `NumericSerialization` types and thus can be processed just like that